### ISSUE TITLE
User defined property as object key instead of system defined 'id' uuid property

### ISSUE THREAD

**Issue by eostis** [weaviate/weaviate#1684]
Hi,

Weaviate requires the  predefined uuid field 'id' to be the key for all objects. It is used for instance to delete or update objects.

But 100% applications have their own primary key (SQL, WordPress, ...), and most often it is not a uuid4 formatted type (no way to copy it to the system 'id' property then).

Each application therefore must generate a uuid4, send it to the API as 'id', and store it in the application on each object to be able to use it later for updates or deletes. I've just done that on my WordPress plugin.

It would much easier for applications if in the Weaviate class schema a property could be marked as 'primary', and use it everywhere in the API where 'id' is currently used (delete, and update at least).

**Comment by etiennedi** [comment]
I absolutely agree, but this would be quite a big change at the moment that is unlikely to be prioritized over more pressing features in the short term. It would most likely also be a breaking change as it would affect how objects are stored. So, to aid with backward compatibility we'd need to automatically migrate data on update - making this even bigger.

Having said that, there are a few workarounds you can do at the moment:

- You can generate uuids based on other ID properties using [UUID v3 or v5](https://en.wikipedia.org/wiki/Universally_unique_identifier#Versions_3_and_5_(namespace_name-based)). The latter uses SHA1 for hashing, so the likelyhood of a hash collision is very, very small. 
- You can always use where filters for lookups by other properites, e.g. to use a custom `myId` field you can set something like `where: {valueString: "123456", path:["myId"], operator: Equal}`. This has some extra cost compared to an exact lookup, but for many use cases the cost shouldn't be too high.

**Comment by eostis** [comment]
I agree this is not a #1 priority: on my side, I built a workaround by mapping (and storing) WordPress ids to Weaviate uuids.

Without this workaround, you can just use Weaviate "out of the box" to insert data. Not to delete, nor update.

**Comment by etiennedi** [comment]
> Without this workaround, you can just use Weaviate "out of the box" to insert data. Not to delete, nor update.

If you use a v3 or v5 UUID generated from the other ID, this should also work during delete and update operations.

**Comment by etiennedi** [comment]
keep open for the long-term

**Comment by etiennedi** [comment]
Keep open.

**Comment by zhenglaizhang** [comment]
> * You can always use where filters for lookups by other properites, e.g. to use a custom `myId` field you can set something like `where: {valueString: "123456", path:["myId"], operator: Equal}`. This has some extra cost compared to an exact lookup, but for many use cases the cost shouldn't be too high.

@etiennedi thanks for the solution, we have the same requirement, and I am trying to see if I can use a customized field to store the id (as int string) and retrieve the object by id. And it seems the perf is not good, my use case requires high throughput and low latency fetching by id. Not sure if it's related to my local env (Mac M1). 
**Curious about what's the extra cost here.**

**Comment by illagrenan** [comment]
My application uses integers as the primary identifier. To avoid having to generate random UUIDs (and remember those), I use this trick (Python simply pads the number with zeros):

```python
import uuid

# INT up to 128 bits
my_app_id = 42

# Convert INT to UUID when travelling to Weaviate:
assert str(uuid.UUID(int=42)) == '00000000-0000-0000-0000-00000000002a'

# And convert the UUID back to INT
assert int(uuid.UUID('00000000-0000-0000-0000-00000000002a')) == 42
```

This method works great. However, it has one drawback. The ID in Weaviate must be globally unique 😥, so it is not possible to store multiple objects with the same ID in Weaviate (even if they are in different classes).

**Comment by etiennedi** [comment]
@illagrenan 

This is a great idea and workaround

> The ID in Weaviate must be globally unique 😥, so it is not possible to store multiple objects with the same ID in Weaviate (even if they are in different classes).

We are aware of this and looking for better solutions. Ideally, the class really acts as a full namespace. The main problem at the moment is that this would be a breaking change as the current beacon format used in cross-references does not include the className, but it would require this for it to be non-breaking.

**Comment by etiennedi** [comment]
This ticket is older than our voting roadmap, but I think it could easily get a lot of upvotes. Added the `backlog` label and reopening ;-)

**Comment by sandstrom** [comment]
Agree, this would be quite useful to us as well. Nice when the authoritative data store is somewhere else. You'd like to use the primary key from another DB.

**Comment by simplynaveen20** [comment]
Recently onboarded weaviate, I found this UUID behavior little odd, feature of allowing to choose our own key instead of UUID would be helpful .
